/*
* @license Apache-2.0
*
* Copyright (c) 2025 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

import zeros = require( './../../../zeros' );
import push = require( './index' );


// TESTS //

// The function returns an ndarray...
{
	const x = zeros( [ 2, 2 ] );

	push( x, 1.0 ); // $ExpectType float64ndarray
}

// The compiler throws an error if the function is provided a first argument which is not an ndarray...
{
	push( '10', 1.0 ); // $ExpectError
	push( 10, 1.0 ); // $ExpectError
	push( false, 1.0 ); // $ExpectError
	push( true, 1.0 ); // $ExpectError
	push( null, 1.0 ); // $ExpectError
	push( [], 1.0 ); // $ExpectError
	push( {}, 1.0 ); // $ExpectError
	push( ( x: number ): number => x, 1.0 ); // $ExpectError
}

// The compiler throws an error if the function is provided an unsupported number of arguments...
{
	push(); // $ExpectError
}

// Attached to the function is an `assign` method which returns an ndarray...
{
	const x = zeros( [ 2 ] );
	const out = zeros( [ 3 ] );

	push.assign( x, 1.0, out ); // $ExpectType float64ndarray
}

// The compiler throws an error if the `assign` method is provided a first argument which is not an ndarray...
{
	const out = zeros( [ 3 ] );

	push.assign( '10', 1.0, out ); // $ExpectError
	push.assign( 10, 1.0, out ); // $ExpectError
	push.assign( false, 1.0, out ); // $ExpectError
	push.assign( true, 1.0, out ); // $ExpectError
	push.assign( null, 1.0, out ); // $ExpectError
	push.assign( [], 1.0, out ); // $ExpectError
	push.assign( {}, 1.0, out ); // $ExpectError
	push.assign( ( x: number ): number => x, 1.0, out ); // $ExpectError
}

// The compiler throws an error if the `assign` method is provided an output argument which is not an ndarray...
{
	const x = zeros( [ 3 ] );

	push.assign( x, 1.0, '10' ); // $ExpectError
	push.assign( x, 1.0, 10 ); // $ExpectError
	push.assign( x, 1.0, false ); // $ExpectError
	push.assign( x, 1.0, true ); // $ExpectError
	push.assign( x, 1.0, null ); // $ExpectError
	push.assign( x, 1.0, [] ); // $ExpectError
	push.assign( x, 1.0, {} ); // $ExpectError
	push.assign( x, 1.0, ( x: number ): number => x ); // $ExpectError
}

// The compiler throws an error if the `assign` method is provided an unsupported number of arguments...
{
	const x = zeros( [ 3 ] );

	push.assign(); // $ExpectError
	push.assign( x ); // $ExpectError
	push.assign( x, 1.0 ); // $ExpectError
}
