/*
* @license Apache-2.0
*
* Copyright (c) 2025 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

import empty = require( './../../../../base/empty' );
import toReversedDimension = require( './index' );


// TESTS //

// The function returns an ndarray...
{
	const order = 'row-major';
	const sh = [ 2, 2 ];

	toReversedDimension( empty( 'float64', sh, order ), 0 ); // $ExpectType float64ndarray
	toReversedDimension( empty( 'float32', sh, order ), 0 ); // $ExpectType float32ndarray
	toReversedDimension( empty( 'complex128', sh, order ), 0 ); // $ExpectType complex128ndarray
	toReversedDimension( empty( 'complex64', sh, order ), 0 ); // $ExpectType complex64ndarray
	toReversedDimension( empty( 'int32', sh, order ), 0 ); // $ExpectType int32ndarray
	toReversedDimension( empty( 'int16', sh, order ), 0 ); // $ExpectType int16ndarray
	toReversedDimension( empty( 'int8', sh, order ), 0 ); // $ExpectType int8ndarray
	toReversedDimension( empty( 'uint32', sh, order ), 0 ); // $ExpectType uint32ndarray
	toReversedDimension( empty( 'uint16', sh, order ), 0 ); // $ExpectType uint16ndarray
	toReversedDimension( empty( 'uint8', sh, order ), 0 ); // $ExpectType uint8ndarray
	toReversedDimension( empty( 'uint8c', sh, order ), 0 ); // $ExpectType uint8cndarray
	toReversedDimension( empty( 'generic', sh, order ), 0 ); // $ExpectType genericndarray<number>
}

// The compiler throws an error if the function is provided a first argument which is not an ndarray...
{
	toReversedDimension( '10', 0 ); // $ExpectError
	toReversedDimension( 10, 0 ); // $ExpectError
	toReversedDimension( false, 0 ); // $ExpectError
	toReversedDimension( true, 0 ); // $ExpectError
	toReversedDimension( null, 0 ); // $ExpectError
	toReversedDimension( [], 0 ); // $ExpectError
	toReversedDimension( {}, 0 ); // $ExpectError
	toReversedDimension( ( x: number ): number => x, 0 ); // $ExpectError
}

// The compiler throws an error if the function is provided a second argument which is not a number...
{
	const x = empty( 'float64', [ 2, 2 ], 'row-major' );

	toReversedDimension( x, '5' ); // $ExpectError
	toReversedDimension( x, true ); // $ExpectError
	toReversedDimension( x, false ); // $ExpectError
	toReversedDimension( x, null ); // $ExpectError
	toReversedDimension( x, undefined ); // $ExpectError
	toReversedDimension( x, [ '5' ] ); // $ExpectError
	toReversedDimension( x, {} ); // $ExpectError
	toReversedDimension( x, ( x: number ): number => x ); // $ExpectError
}

// The compiler throws an error if the function is provided an unsupported number of arguments...
{
	const x = empty( 'float64', [ 2, 2 ], 'row-major' );

	toReversedDimension(); // $ExpectError
	toReversedDimension( x ); // $ExpectError
	toReversedDimension( x, 0, {} ); // $ExpectError
}
