/*
* @license Apache-2.0
*
* Copyright (c) 2026 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

import zeros = require( './../../../../zeros' );
import atleastnd = require( './index' );


// TESTS //

// The function returns an array of ndarrays...
{
	const x = zeros( [ 2, 2 ] );
	const y = zeros( [ 2, 2, 2 ] );

	atleastnd( 3, [ x ] ); // $ExpectType ndarray[]
	atleastnd( 3, [ x, y ] ); // $ExpectType ndarray[]
	atleastnd( 3, [ x, y, x ] ); // $ExpectType ndarray[]
}

// The compiler throws an error if the function is provided a first argument which is not a number...
{
	const x = zeros( [ 2, 2 ] );
	const y = zeros( [ 2, 2, 2 ] );

	atleastnd( '5', [ x, y ] ); // $ExpectError
	atleastnd( true, [ x, y ] ); // $ExpectError
	atleastnd( false, [ x, y ] ); // $ExpectError
	atleastnd( null, [ x, y ] ); // $ExpectError
	atleastnd( {}, [ x, y ] ); // $ExpectError
	atleastnd( [ '5' ], [ x, y ] ); // $ExpectError
	atleastnd( ( x: number ): number => x, [ x, y ] ); // $ExpectError
}

// The compiler throws an error if the function is provided a second argument which is not an array ...
{
	atleastnd( 3, 5 ); // $ExpectError
	atleastnd( 3, true ); // $ExpectError
	atleastnd( 3, false ); // $ExpectError
	atleastnd( 3, null ); // $ExpectError
	atleastnd( 3, {} ); // $ExpectError
	atleastnd( 3, ( x: number ): number => x ); // $ExpectError
}

// The compiler throws an error if the function is provided an unsupported number of arguments...
{
	const x = zeros( [ 2, 2 ] );
	const y = zeros( [ 2, 2, 2 ] );

	atleastnd(); // $ExpectError
	atleastnd( 3 ); // $ExpectError
	atleastnd( 3, [ x, y ], {} ); // $ExpectError
}
