/*
* @license Apache-2.0
*
* Copyright (c) 2026 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

import empty = require( './../../../base/empty' );
import spreadDimensions = require( './index' );


// TESTS //

// The function returns an ndarray...
{
	const order = 'row-major';
	const sh = [ 2, 2 ];

	spreadDimensions( 3, empty( 'float64', sh, order ), [ 0, 1 ] ); // $ExpectType float64ndarray
	spreadDimensions( 5, empty( 'complex64', sh, order ), [ 1, 3 ] ); // $ExpectType complex64ndarray
	spreadDimensions( 4, empty( 'int32', sh, order ), [ -2, -1 ] ); // $ExpectType int32ndarray
}

// The compiler throws an error if the function is provided a first argument which is not a number...
{
	const x = empty( 'float64', [ 2, 2 ], 'row-major' );

	spreadDimensions( '5', x, [ 0, 1 ] ); // $ExpectError
	spreadDimensions( false, x, [ 0, 1 ] ); // $ExpectError
	spreadDimensions( true, x, [ 0, 1 ] ); // $ExpectError
	spreadDimensions( null, x, [ 0, 1 ] ); // $ExpectError
	spreadDimensions( [], x, [ 0, 1 ] ); // $ExpectError
	spreadDimensions( {}, x, [ 0, 1 ] ); // $ExpectError
	spreadDimensions( ( x: number ): number => x, x, [ 0, 1 ] ); // $ExpectError
}

// The compiler throws an error if the function is provided a second argument which is not an ndarray...
{
	spreadDimensions( 5, '10', [ 0, 1 ] ); // $ExpectError
	spreadDimensions( 5, 10, [ 0, 1 ] ); // $ExpectError
	spreadDimensions( 5, false, [ 0, 1 ] ); // $ExpectError
	spreadDimensions( 5, true, [ 0, 1 ] ); // $ExpectError
	spreadDimensions( 5, null, [ 0, 1 ] ); // $ExpectError
	spreadDimensions( 5, [], [ 0, 1 ] ); // $ExpectError
	spreadDimensions( 5, {}, [ 0, 1 ] ); // $ExpectError
	spreadDimensions( 5, ( x: number ): number => x, [ 0, 1 ] ); // $ExpectError
}

// The compiler throws an error if the function is provided a third argument which is not a collection of numbers...
{
	const x = empty( 'float64', [ 2, 2 ], 'row-major' );

	spreadDimensions( 5, x, '1' ); // $ExpectError
	spreadDimensions( 5, x, 1 ); // $ExpectError
	spreadDimensions( 5, x, false ); // $ExpectError
	spreadDimensions( 5, x, true ); // $ExpectError
	spreadDimensions( 5, x, null ); // $ExpectError
	spreadDimensions( 5, x, {} ); // $ExpectError
	spreadDimensions( 5, x, ( x: number ): number => x ); // $ExpectError
}

// The compiler throws an error if the function is provided an unsupported number of arguments...
{
	const x = empty( 'float64', [ 2, 2 ], 'row-major' );

	spreadDimensions(); // $ExpectError
	spreadDimensions( 5 ); // $ExpectError
	spreadDimensions( 5, x ); // $ExpectError
	spreadDimensions( 5, x, [ 0, 1 ], {} ); // $ExpectError
}
