/*
* @license Apache-2.0
*
* Copyright (c) 2025 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

/// <reference types="@stdlib/types"/>

import zeros = require( './../../../zeros' );
import concat1d = require( './index' );


// TESTS //

// The function returns an ndarray...
{
	const x = zeros( [ 2, 2 ] );
	const y = zeros( [ 2, 3 ] );

	concat1d( [ x, y ] ); // $ExpectType typedndarray<number>
	concat1d( x, y ); // $ExpectType typedndarray<number>
}

// The compiler throws an error if the function is provided an unsupported number of arguments...
{
	concat1d(); // $ExpectError
}

// Attached to the function is an `assign` method which returns an ndarray...
{
	const x = zeros( [ 2, 2 ] );
	const y = zeros( [ 2, 3 ] );
	const out = zeros( [ 2, 5 ] );

	concat1d.assign( [ x, y ], out ); // $ExpectType float64ndarray
	concat1d.assign( x, y, out ); // $ExpectType float64ndarray
}

// The compiler throws an error if the function is provided an unsupported number of arguments...
{
	const x = zeros( [ 2, 2 ] );

	concat1d.assign(); // $ExpectError
	concat1d.assign( x ); // $ExpectError
}
