/*
* @license Apache-2.0
*
* Copyright (c) 2025 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

import zeros = require( './../../../../zeros' );
import scalar2ndarray = require( './../../../../from-scalar' );
import some = require( './index' );


// TESTS //

// The function returns a boolean...
{
	const x = zeros( [ 2, 2 ] );
	const n = scalar2ndarray( 2, { 'dtype': 'generic' } );

	some( [ x, n ] ); // $ExpectType boolean
}

// The compiler throws an error if the function is provided a first argument which is not an array-like object containing ndarray-like objects...
{
	some( 5 ); // $ExpectError
	some( true ); // $ExpectError
	some( false ); // $ExpectError
	some( null ); // $ExpectError
	some( undefined ); // $ExpectError
	some( {} ); // $ExpectError
	some( [ 1 ] ); // $ExpectError
	some( ( x: number ): number => x ); // $ExpectError

	some( 5 ); // $ExpectError
	some( true ); // $ExpectError
	some( false ); // $ExpectError
	some( null ); // $ExpectError
	some( undefined ); // $ExpectError
	some( {} ); // $ExpectError
	some( [ 1 ] ); // $ExpectError
	some( ( x: number ): number => x ); // $ExpectError
}

// The compiler throws an error if the function is provided an unsupported number of arguments...
{
	const x = zeros( [ 2, 2 ] );
	const n = scalar2ndarray( 2, { 'dtype': 'generic' } );

	some(); // $ExpectError
	some( [ x ] ); // $ExpectError
	some( [ x, n ], {}, {} ); // $ExpectError
}
