/*
* @license Apache-2.0
*
* Copyright (c) 2026 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

/// <reference types="@stdlib/types"/>

import zeros = require( './../../../../zeros' );
import maybeBroadcastArrayExceptDimensions = require( './index' );


// TESTS //

// The function returns an ndarray...
{
	const x = zeros( [ 2, 2 ] );

	maybeBroadcastArrayExceptDimensions( x, [ 2, 2, 2 ], [ -2 ] ); // $ExpectType ndarray
}

// The compiler throws an error if the function is not provided a first argument which is an ndarray...
{
	maybeBroadcastArrayExceptDimensions( '5', [ 2, 2, 2 ], [ -2 ] ); // $ExpectError
	maybeBroadcastArrayExceptDimensions( 5, [ 2, 2, 2 ], [ -2 ] ); // $ExpectError
	maybeBroadcastArrayExceptDimensions( true, [ 2, 2, 2 ], [ -2 ] ); // $ExpectError
	maybeBroadcastArrayExceptDimensions( false, [ 2, 2, 2 ], [ -2 ] ); // $ExpectError
	maybeBroadcastArrayExceptDimensions( null, [ 2, 2, 2 ], [ -2 ] ); // $ExpectError
	maybeBroadcastArrayExceptDimensions( {}, [ 2, 2, 2 ], [ -2 ] ); // $ExpectError
	maybeBroadcastArrayExceptDimensions( [ '5' ], [ 2, 2, 2 ], [ -2 ] ); // $ExpectError
	maybeBroadcastArrayExceptDimensions( ( x: number ): number => x, [ 2, 2, 2 ], [ -2 ] ); // $ExpectError
}

// The compiler throws an error if the function is not provided a second argument which is an array-like object containing numbers...
{
	const x = zeros( [ 2, 2 ] );

	maybeBroadcastArrayExceptDimensions( x, '5', [ -2 ] ); // $ExpectError
	maybeBroadcastArrayExceptDimensions( x, 5, [ -2 ] ); // $ExpectError
	maybeBroadcastArrayExceptDimensions( x, true, [ -2 ] ); // $ExpectError
	maybeBroadcastArrayExceptDimensions( x, false, [ -2 ] ); // $ExpectError
	maybeBroadcastArrayExceptDimensions( x, null, [ -2 ] ); // $ExpectError
	maybeBroadcastArrayExceptDimensions( x, {}, [ -2 ] ); // $ExpectError
	maybeBroadcastArrayExceptDimensions( x, [ '5' ], [ -2 ] ); // $ExpectError
	maybeBroadcastArrayExceptDimensions( x, ( x: number ): number => x, [ -2 ] ); // $ExpectError
}

// The compiler throws an error if the function is not provided a third argument which is not an array-like object containing numbers...
{
	const x = zeros( [ 2, 2 ] );

	maybeBroadcastArrayExceptDimensions( x, [ 2, 2, 2 ], '5' ); // $ExpectError
	maybeBroadcastArrayExceptDimensions( x, [ 2, 2, 2 ], 5 ); // $ExpectError
	maybeBroadcastArrayExceptDimensions( x, [ 2, 2, 2 ], true ); // $ExpectError
	maybeBroadcastArrayExceptDimensions( x, [ 2, 2, 2 ], false ); // $ExpectError
	maybeBroadcastArrayExceptDimensions( x, [ 2, 2, 2 ], null ); // $ExpectError
	maybeBroadcastArrayExceptDimensions( x, [ 2, 2, 2 ], [ '5' ] ); // $ExpectError
	maybeBroadcastArrayExceptDimensions( x, [ 2, 2, 2 ], ( x: number ): number => x ); // $ExpectError
}

// The compiler throws an error if the function is provided an unsupported number of arguments...
{
	const x = zeros( [ 2, 2 ] );

	maybeBroadcastArrayExceptDimensions(); // $ExpectError
	maybeBroadcastArrayExceptDimensions( x ); // $ExpectError
	maybeBroadcastArrayExceptDimensions( x, [ 1, 2, 3 ], [ -2 ], {} ); // $ExpectError
}
