/*
* @license Apache-2.0
*
* Copyright (c) 2025 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

import zeros = require( './../../../../base/zeros' );
import empty = require( './../../../../base/empty' );
import copy = require( './index' );


// TESTS //

// The function returns an ndarray...
{
	const sh = [ 2, 2 ];
	const ord = 'row-major';

	copy( zeros( 'float64', sh, ord ) ); // $ExpectType float64ndarray
	copy( zeros( 'float32', sh, ord ) ); // $ExpectType float32ndarray
	copy( zeros( 'complex128', sh, ord ) ); // $ExpectType complex128ndarray
	copy( zeros( 'complex64', sh, ord ) ); // $ExpectType complex64ndarray
	copy( zeros( 'int32', sh, ord ) ); // $ExpectType int32ndarray
	copy( zeros( 'int16', sh, ord ) ); // $ExpectType int16ndarray
	copy( zeros( 'int8', sh, ord ) ); // $ExpectType int8ndarray
	copy( zeros( 'uint32', sh, ord ) ); // $ExpectType uint32ndarray
	copy( zeros( 'uint16', sh, ord ) ); // $ExpectType uint16ndarray
	copy( zeros( 'uint8', sh, ord ) ); // $ExpectType uint8ndarray
	copy( zeros( 'uint8c', sh, ord ) ); // $ExpectType uint8cndarray
	copy( empty( 'bool', sh, ord ) ); // $ExpectType boolndarray
	copy( zeros( 'generic', sh, ord ) ); // $ExpectType genericndarray<number>
}

// The compiler throws an error if the function is provided a first argument is not an ndarray...
{
	copy( '10' ); // $ExpectError
	copy( 10 ); // $ExpectError
	copy( false ); // $ExpectError
	copy( true ); // $ExpectError
	copy( null ); // $ExpectError
	copy( [] ); // $ExpectError
	copy( {} ); // $ExpectError
	copy( ( x: number ): number => x ); // $ExpectError
}

// The compiler throws an error if the function is provided an unsupported number of arguments...
{
	copy(); // $ExpectError
	copy( zeros( 'float64', [ 2, 2 ], 'row-major' ), 1 ); // $ExpectError
}
