/*
* @license Apache-2.0
*
* Copyright (c) 2025 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

import empty = require( './../../../base/empty' );
import zeros = require( './../../../base/zeros' );
import flattenFromBy = require( './index' );

/**
* Evaluates the identity function.
*
* @param x - input value
* @returns input value
*/
function identity( x: any ): any {
	return x;
}

// TESTS //

// The function returns an ndarray...
{
	const sh = [ 2, 2, 2 ];
	const ord = 'row-major';

	flattenFromBy( zeros( 'float64', sh, ord ), 0, identity ); // $ExpectType float64ndarray
	flattenFromBy( zeros( 'float64', sh, ord ), 1, {}, identity ); // $ExpectType float64ndarray
	flattenFromBy( zeros( 'float64', sh, ord ), 0, identity, {} ); // $ExpectType float64ndarray
	flattenFromBy( zeros( 'float64', sh, ord ), 1, {}, identity, {} ); // $ExpectType float64ndarray
	flattenFromBy( zeros( 'float32', sh, ord ), 0, identity ); // $ExpectType float32ndarray
	flattenFromBy( zeros( 'float32', sh, ord ), 1, {}, identity ); // $ExpectType float32ndarray
	flattenFromBy( zeros( 'float32', sh, ord ), 0, identity, {} ); // $ExpectType float32ndarray
	flattenFromBy( zeros( 'float32', sh, ord ), 1, {}, identity, {} ); // $ExpectType float32ndarray
	flattenFromBy( zeros( 'complex64', sh, ord ), 0, identity ); // $ExpectType complex64ndarray
	flattenFromBy( zeros( 'complex64', sh, ord ), 1, {}, identity ); // $ExpectType complex64ndarray
	flattenFromBy( zeros( 'complex64', sh, ord ), 0, identity, {} ); // $ExpectType complex64ndarray
	flattenFromBy( zeros( 'complex64', sh, ord ), 1, {}, identity, {} ); // $ExpectType complex64ndarray
	flattenFromBy( zeros( 'complex128', sh, ord ), 0, identity ); // $ExpectType complex128ndarray
	flattenFromBy( zeros( 'complex128', sh, ord ), 1, {}, identity ); // $ExpectType complex128ndarray
	flattenFromBy( zeros( 'complex128', sh, ord ), 0, identity, {} ); // $ExpectType complex128ndarray
	flattenFromBy( zeros( 'complex128', sh, ord ), 1, {}, identity, {} ); // $ExpectType complex128ndarray
	flattenFromBy( zeros( 'int32', sh, ord ), 0, identity ); // $ExpectType int32ndarray
	flattenFromBy( zeros( 'int32', sh, ord ), 1, {}, identity ); // $ExpectType int32ndarray
	flattenFromBy( zeros( 'int32', sh, ord ), 0, identity, {} ); // $ExpectType int32ndarray
	flattenFromBy( zeros( 'int32', sh, ord ), 1, {}, identity, {} ); // $ExpectType int32ndarray
	flattenFromBy( zeros( 'int16', sh, ord ), 0, identity ); // $ExpectType int16ndarray
	flattenFromBy( zeros( 'int16', sh, ord ), 1, {}, identity ); // $ExpectType int16ndarray
	flattenFromBy( zeros( 'int16', sh, ord ), 0, identity, {} ); // $ExpectType int16ndarray
	flattenFromBy( zeros( 'int16', sh, ord ), 1, {}, identity, {} ); // $ExpectType int16ndarray
	flattenFromBy( zeros( 'int8', sh, ord ), 0, identity ); // $ExpectType int8ndarray
	flattenFromBy( zeros( 'int8', sh, ord ), 1, {}, identity ); // $ExpectType int8ndarray
	flattenFromBy( zeros( 'int8', sh, ord ), 0, identity, {} ); // $ExpectType int8ndarray
	flattenFromBy( zeros( 'int8', sh, ord ), 1, {}, identity, {} ); // $ExpectType int8ndarray
	flattenFromBy( zeros( 'uint32', sh, ord ), 0, identity ); // $ExpectType uint32ndarray
	flattenFromBy( zeros( 'uint32', sh, ord ), 1, {}, identity ); // $ExpectType uint32ndarray
	flattenFromBy( zeros( 'uint32', sh, ord ), 0, identity, {} ); // $ExpectType uint32ndarray
	flattenFromBy( zeros( 'uint32', sh, ord ), 1, {}, identity, {} ); // $ExpectType uint32ndarray
	flattenFromBy( zeros( 'uint16', sh, ord ), 0, identity ); // $ExpectType uint16ndarray
	flattenFromBy( zeros( 'uint16', sh, ord ), 1, {}, identity ); // $ExpectType uint16ndarray
	flattenFromBy( zeros( 'uint16', sh, ord ), 0, identity, {} ); // $ExpectType uint16ndarray
	flattenFromBy( zeros( 'uint16', sh, ord ), 1, {}, identity, {} ); // $ExpectType uint16ndarray
	flattenFromBy( zeros( 'uint8', sh, ord ), 0, identity ); // $ExpectType uint8ndarray
	flattenFromBy( zeros( 'uint8', sh, ord ), 1, {}, identity ); // $ExpectType uint8ndarray
	flattenFromBy( zeros( 'uint8', sh, ord ), 0, identity, {} ); // $ExpectType uint8ndarray
	flattenFromBy( zeros( 'uint8', sh, ord ), 1, {}, identity, {} ); // $ExpectType uint8ndarray
	flattenFromBy( zeros( 'uint8c', sh, ord ), 0, identity ); // $ExpectType uint8cndarray
	flattenFromBy( zeros( 'uint8c', sh, ord ), 1, {}, identity ); // $ExpectType uint8cndarray
	flattenFromBy( zeros( 'uint8c', sh, ord ), 0, identity, {} ); // $ExpectType uint8cndarray
	flattenFromBy( zeros( 'uint8c', sh, ord ), 1, {}, identity, {} ); // $ExpectType uint8cndarray
	flattenFromBy( empty( 'bool', sh, ord ), 0, identity ); // $ExpectType boolndarray
	flattenFromBy( empty( 'bool', sh, ord ), 1, {}, identity ); // $ExpectType boolndarray
	flattenFromBy( empty( 'bool', sh, ord ), 0, identity, {} ); // $ExpectType boolndarray
	flattenFromBy( empty( 'bool', sh, ord ), 1, {}, identity, {} ); // $ExpectType boolndarray
	flattenFromBy( zeros( 'generic', sh, ord ), 0, identity ); // $ExpectType genericndarray<number>
	flattenFromBy( zeros( 'generic', sh, ord ), 1, {}, identity ); // $ExpectType genericndarray<number>
	flattenFromBy( zeros( 'generic', sh, ord ), 0, identity, {} ); // $ExpectType genericndarray<number>
	flattenFromBy( zeros( 'generic', sh, ord ), 1, {}, identity, {} ); // $ExpectType genericndarray<number>

	flattenFromBy( zeros( 'float64', sh, ord ), 0, { 'dtype': 'float32' }, identity ); // $ExpectType float32ndarray
	flattenFromBy( zeros( 'float64', sh, ord ), 1, { 'dtype': 'generic' }, identity ); // $ExpectType genericndarray<any>
	flattenFromBy( zeros( 'generic', sh, ord ), 0, { 'dtype': 'float64' }, identity ); // $ExpectType float64ndarray
	flattenFromBy( zeros( 'generic', sh, ord ), 1, { 'dtype': 'generic' }, identity ); // $ExpectType genericndarray<any>
}

// The compiler throws an error if the function is provided a first argument which is not an ndarray...
{
	flattenFromBy( '5', 0, identity ); // $ExpectError
	flattenFromBy( 5, 0, identity ); // $ExpectError
	flattenFromBy( true, 0, identity ); // $ExpectError
	flattenFromBy( false, 0, identity ); // $ExpectError
	flattenFromBy( null, 0, identity ); // $ExpectError
	flattenFromBy( undefined, 0, identity ); // $ExpectError
	flattenFromBy( {}, 0, identity ); // $ExpectError
	flattenFromBy( [ 1 ], 0, identity ); // $ExpectError
	flattenFromBy( ( x: number ): number => x, 0, identity ); // $ExpectError

	flattenFromBy( '5', 0, {}, identity ); // $ExpectError
	flattenFromBy( 5, 0, {}, identity ); // $ExpectError
	flattenFromBy( true, 0, {}, identity ); // $ExpectError
	flattenFromBy( false, 0, {}, identity ); // $ExpectError
	flattenFromBy( null, 0, {}, identity ); // $ExpectError
	flattenFromBy( undefined, 0, {}, identity ); // $ExpectError
	flattenFromBy( {}, 0, {}, identity ); // $ExpectError
	flattenFromBy( [ 1 ], 0, {}, identity ); // $ExpectError
	flattenFromBy( ( x: number ): number => x, 0, {}, identity ); // $ExpectError

	flattenFromBy( '5', 0, identity, {} ); // $ExpectError
	flattenFromBy( 5, 0, identity, {} ); // $ExpectError
	flattenFromBy( true, 0, identity, {} ); // $ExpectError
	flattenFromBy( false, 0, identity, {} ); // $ExpectError
	flattenFromBy( null, 0, identity, {} ); // $ExpectError
	flattenFromBy( undefined, 0, identity, {} ); // $ExpectError
	flattenFromBy( {}, 0, identity, {} ); // $ExpectError
	flattenFromBy( [ 1 ], 0, identity, {} ); // $ExpectError
	flattenFromBy( ( x: number ): number => x, 0, identity, {} ); // $ExpectError

	flattenFromBy( '5', 0, {}, identity, {} ); // $ExpectError
	flattenFromBy( 5, 0, {}, identity, {} ); // $ExpectError
	flattenFromBy( true, 0, {}, identity, {} ); // $ExpectError
	flattenFromBy( false, 0, {}, identity, {} ); // $ExpectError
	flattenFromBy( null, 0, {}, identity, {} ); // $ExpectError
	flattenFromBy( undefined, 0, {}, identity, {} ); // $ExpectError
	flattenFromBy( {}, 0, {}, identity, {} ); // $ExpectError
	flattenFromBy( [ 1 ], 0, {}, identity, {} ); // $ExpectError
	flattenFromBy( ( x: number ): number => x, 0, {}, identity, {} ); // $ExpectError
}

// The compiler throws an error if the function is provided a second argument which is not a number...
{
	flattenFromBy( zeros( 'float64', [ 2, 2, 2 ], 'row-major' ), '5', identity ); // $ExpectError
	flattenFromBy( zeros( 'float64', [ 2, 2, 2 ], 'row-major' ), true, identity ); // $ExpectError
	flattenFromBy( zeros( 'float64', [ 2, 2, 2 ], 'row-major' ), false, identity ); // $ExpectError
	flattenFromBy( zeros( 'float64', [ 2, 2, 2 ], 'row-major' ), null, identity ); // $ExpectError
	flattenFromBy( zeros( 'float64', [ 2, 2, 2 ], 'row-major' ), [ 1 ], identity ); // $ExpectError
	flattenFromBy( zeros( 'float64', [ 2, 2, 2 ], 'row-major' ), ( x: number ): number => x, identity ); // $ExpectError

	flattenFromBy( zeros( 'complex128', [ 2, 2, 2 ], 'row-major' ), '5', identity ); // $ExpectError
	flattenFromBy( zeros( 'complex128', [ 2, 2, 2 ], 'row-major' ), true, identity ); // $ExpectError
	flattenFromBy( zeros( 'complex128', [ 2, 2, 2 ], 'row-major' ), false, identity ); // $ExpectError
	flattenFromBy( zeros( 'complex128', [ 2, 2, 2 ], 'row-major' ), null, identity ); // $ExpectError
	flattenFromBy( zeros( 'complex128', [ 2, 2, 2 ], 'row-major' ), [ 1 ], identity ); // $ExpectError
	flattenFromBy( zeros( 'complex128', [ 2, 2, 2 ], 'row-major' ), ( x: number ): number => x, identity ); // $ExpectError

	flattenFromBy( zeros( 'generic', [ 2, 2, 2 ], 'row-major' ), '5', identity ); // $ExpectError
	flattenFromBy( zeros( 'generic', [ 2, 2, 2 ], 'row-major' ), true, identity ); // $ExpectError
	flattenFromBy( zeros( 'generic', [ 2, 2, 2 ], 'row-major' ), false, identity ); // $ExpectError
	flattenFromBy( zeros( 'generic', [ 2, 2, 2 ], 'row-major' ), null, identity ); // $ExpectError
	flattenFromBy( zeros( 'generic', [ 2, 2, 2 ], 'row-major' ), [ 1 ], identity ); // $ExpectError
	flattenFromBy( zeros( 'generic', [ 2, 2, 2 ], 'row-major' ), ( x: number ): number => x, identity ); // $ExpectError

	flattenFromBy( zeros( 'float64', [ 2, 2, 2 ], 'row-major' ), '5', {}, identity ); // $ExpectError
	flattenFromBy( zeros( 'float64', [ 2, 2, 2 ], 'row-major' ), true, {}, identity ); // $ExpectError
	flattenFromBy( zeros( 'float64', [ 2, 2, 2 ], 'row-major' ), false, {}, identity ); // $ExpectError
	flattenFromBy( zeros( 'float64', [ 2, 2, 2 ], 'row-major' ), null, {}, identity ); // $ExpectError
	flattenFromBy( zeros( 'float64', [ 2, 2, 2 ], 'row-major' ), [ 1 ], {}, identity ); // $ExpectError
	flattenFromBy( zeros( 'float64', [ 2, 2, 2 ], 'row-major' ), ( x: number ): number => x, {}, identity ); // $ExpectError

	flattenFromBy( zeros( 'complex128', [ 2, 2, 2 ], 'row-major' ), '5', {}, identity ); // $ExpectError
	flattenFromBy( zeros( 'complex128', [ 2, 2, 2 ], 'row-major' ), true, {}, identity ); // $ExpectError
	flattenFromBy( zeros( 'complex128', [ 2, 2, 2 ], 'row-major' ), false, {}, identity ); // $ExpectError
	flattenFromBy( zeros( 'complex128', [ 2, 2, 2 ], 'row-major' ), null, {}, identity ); // $ExpectError
	flattenFromBy( zeros( 'complex128', [ 2, 2, 2 ], 'row-major' ), [ 1 ], {}, identity ); // $ExpectError
	flattenFromBy( zeros( 'complex128', [ 2, 2, 2 ], 'row-major' ), ( x: number ): number => x, {}, identity ); // $ExpectError

	flattenFromBy( zeros( 'generic', [ 2, 2, 2 ], 'row-major' ), '5', {}, identity ); // $ExpectError
	flattenFromBy( zeros( 'generic', [ 2, 2, 2 ], 'row-major' ), true, {}, identity ); // $ExpectError
	flattenFromBy( zeros( 'generic', [ 2, 2, 2 ], 'row-major' ), false, {}, identity ); // $ExpectError
	flattenFromBy( zeros( 'generic', [ 2, 2, 2 ], 'row-major' ), null, {}, identity ); // $ExpectError
	flattenFromBy( zeros( 'generic', [ 2, 2, 2 ], 'row-major' ), [ 1 ], {}, identity ); // $ExpectError
	flattenFromBy( zeros( 'generic', [ 2, 2, 2 ], 'row-major' ), ( x: number ): number => x, {}, identity ); // $ExpectError
}

// The compiler throws an error if the function is provided an options argument which is not an object...
{
	flattenFromBy( zeros( 'float64', [ 2, 2, 2 ], 'row-major' ), 0, '5', identity ); // $ExpectError
	flattenFromBy( zeros( 'float64', [ 2, 2, 2 ], 'row-major' ), 0, true, identity ); // $ExpectError
	flattenFromBy( zeros( 'float64', [ 2, 2, 2 ], 'row-major' ), 0, false, identity ); // $ExpectError
	flattenFromBy( zeros( 'float64', [ 2, 2, 2 ], 'row-major' ), 0, null, identity ); // $ExpectError
	flattenFromBy( zeros( 'float64', [ 2, 2, 2 ], 'row-major' ), 0, [ 1 ], identity ); // $ExpectError

	flattenFromBy( zeros( 'complex128', [ 2, 2, 2 ], 'row-major' ), 0, '5', identity ); // $ExpectError
	flattenFromBy( zeros( 'complex128', [ 2, 2, 2 ], 'row-major' ), 0, true, identity ); // $ExpectError
	flattenFromBy( zeros( 'complex128', [ 2, 2, 2 ], 'row-major' ), 0, false, identity ); // $ExpectError
	flattenFromBy( zeros( 'complex128', [ 2, 2, 2 ], 'row-major' ), 0, null, identity ); // $ExpectError
	flattenFromBy( zeros( 'complex128', [ 2, 2, 2 ], 'row-major' ), 0, [ 1 ], identity ); // $ExpectError

	flattenFromBy( zeros( 'generic', [ 2, 2, 2 ], 'row-major' ), 0, '5', identity ); // $ExpectError
	flattenFromBy( zeros( 'generic', [ 2, 2, 2 ], 'row-major' ), 0, true, identity ); // $ExpectError
	flattenFromBy( zeros( 'generic', [ 2, 2, 2 ], 'row-major' ), 0, false, identity ); // $ExpectError
	flattenFromBy( zeros( 'generic', [ 2, 2, 2 ], 'row-major' ), 0, null, identity ); // $ExpectError
	flattenFromBy( zeros( 'generic', [ 2, 2, 2 ], 'row-major' ), 0, [ 1 ], identity ); // $ExpectError
}

// The compiler throws an error if the function is provided an invalid `order` option...
{
	flattenFromBy( zeros( 'float64', [ 2, 2, 2 ], 'row-major' ), 0, { 'order': '5' }, identity ); // $ExpectError
	flattenFromBy( zeros( 'float64', [ 2, 2, 2 ], 'row-major' ), 0, { 'order': true }, identity ); // $ExpectError
	flattenFromBy( zeros( 'float64', [ 2, 2, 2 ], 'row-major' ), 0, { 'order': false }, identity ); // $ExpectError
	flattenFromBy( zeros( 'float64', [ 2, 2, 2 ], 'row-major' ), 0, { 'order': null }, identity ); // $ExpectError
	flattenFromBy( zeros( 'float64', [ 2, 2, 2 ], 'row-major' ), 0, { 'order': [ 1 ] }, identity ); // $ExpectError
	flattenFromBy( zeros( 'float64', [ 2, 2, 2 ], 'row-major' ), 0, { 'order': ( x: number ): number => x }, identity ); // $ExpectError

	flattenFromBy( zeros( 'complex128', [ 2, 2, 2 ], 'row-major' ), 0, { 'order': '5' }, identity ); // $ExpectError
	flattenFromBy( zeros( 'complex128', [ 2, 2, 2 ], 'row-major' ), 0, { 'order': true }, identity ); // $ExpectError
	flattenFromBy( zeros( 'complex128', [ 2, 2, 2 ], 'row-major' ), 0, { 'order': false }, identity ); // $ExpectError
	flattenFromBy( zeros( 'complex128', [ 2, 2, 2 ], 'row-major' ), 0, { 'order': null }, identity ); // $ExpectError
	flattenFromBy( zeros( 'complex128', [ 2, 2, 2 ], 'row-major' ), 0, { 'order': [ 1 ] }, identity ); // $ExpectError
	flattenFromBy( zeros( 'complex128', [ 2, 2, 2 ], 'row-major' ), 0, { 'order': ( x: number ): number => x }, identity ); // $ExpectError

	flattenFromBy( zeros( 'generic', [ 2, 2, 2 ], 'row-major' ), 0, { 'order': '5' }, identity ); // $ExpectError
	flattenFromBy( zeros( 'generic', [ 2, 2, 2 ], 'row-major' ), 0, { 'order': true }, identity ); // $ExpectError
	flattenFromBy( zeros( 'generic', [ 2, 2, 2 ], 'row-major' ), 0, { 'order': false }, identity ); // $ExpectError
	flattenFromBy( zeros( 'generic', [ 2, 2, 2 ], 'row-major' ), 0, { 'order': null }, identity ); // $ExpectError
	flattenFromBy( zeros( 'generic', [ 2, 2, 2 ], 'row-major' ), 0, { 'order': [ 1 ] }, identity ); // $ExpectError
	flattenFromBy( zeros( 'generic', [ 2, 2, 2 ], 'row-major' ), 0, { 'order': ( x: number ): number => x }, identity ); // $ExpectError
}

// The compiler throws an error if the function is provided an invalid `dtype` option...
{
	flattenFromBy( zeros( 'float64', [ 2, 2, 2 ], 'row-major' ), 0, { 'dtype': '5' }, identity ); // $ExpectError
	flattenFromBy( zeros( 'float64', [ 2, 2, 2 ], 'row-major' ), 0, { 'dtype': true }, identity ); // $ExpectError
	flattenFromBy( zeros( 'float64', [ 2, 2, 2 ], 'row-major' ), 0, { 'dtype': false }, identity ); // $ExpectError
	flattenFromBy( zeros( 'float64', [ 2, 2, 2 ], 'row-major' ), 0, { 'dtype': null }, identity ); // $ExpectError
	flattenFromBy( zeros( 'float64', [ 2, 2, 2 ], 'row-major' ), 0, { 'dtype': [ 1 ] }, identity ); // $ExpectError
	flattenFromBy( zeros( 'float64', [ 2, 2, 2 ], 'row-major' ), 0, { 'dtype': ( x: number ): number => x }, identity ); // $ExpectError

	flattenFromBy( zeros( 'complex128', [ 2, 2, 2 ], 'row-major' ), 0, { 'dtype': '5' }, identity ); // $ExpectError
	flattenFromBy( zeros( 'complex128', [ 2, 2, 2 ], 'row-major' ), 0, { 'dtype': true }, identity ); // $ExpectError
	flattenFromBy( zeros( 'complex128', [ 2, 2, 2 ], 'row-major' ), 0, { 'dtype': false }, identity ); // $ExpectError
	flattenFromBy( zeros( 'complex128', [ 2, 2, 2 ], 'row-major' ), 0, { 'dtype': null }, identity ); // $ExpectError
	flattenFromBy( zeros( 'complex128', [ 2, 2, 2 ], 'row-major' ), 0, { 'dtype': [ 1 ] }, identity ); // $ExpectError
	flattenFromBy( zeros( 'complex128', [ 2, 2, 2 ], 'row-major' ), 0, { 'dtype': ( x: number ): number => x }, identity ); // $ExpectError

	flattenFromBy( zeros( 'generic', [ 2, 2, 2 ], 'row-major' ), 0, { 'dtype': '5' }, identity ); // $ExpectError
	flattenFromBy( zeros( 'generic', [ 2, 2, 2 ], 'row-major' ), 0, { 'dtype': true }, identity ); // $ExpectError
	flattenFromBy( zeros( 'generic', [ 2, 2, 2 ], 'row-major' ), 0, { 'dtype': false }, identity ); // $ExpectError
	flattenFromBy( zeros( 'generic', [ 2, 2, 2 ], 'row-major' ), 0, { 'dtype': null }, identity ); // $ExpectError
	flattenFromBy( zeros( 'generic', [ 2, 2, 2 ], 'row-major' ), 0, { 'dtype': [ 1 ] }, identity ); // $ExpectError
	flattenFromBy( zeros( 'generic', [ 2, 2, 2 ], 'row-major' ), 0, { 'dtype': ( x: number ): number => x }, identity ); // $ExpectError
}

// The compiler throws an error if the function is provided a callback which is not a function...
{
	const x = zeros( 'generic', [ 2, 2 ], 'row-major' );

	flattenFromBy( x, 0, '5' ); // $ExpectError
	flattenFromBy( x, 0, true ); // $ExpectError
	flattenFromBy( x, 0, false ); // $ExpectError
	flattenFromBy( x, 0, null ); // $ExpectError
	flattenFromBy( x, 0, undefined ); // $ExpectError
	flattenFromBy( x, 0, {} ); // $ExpectError
	flattenFromBy( x, 0, [ 1 ] ); // $ExpectError

	flattenFromBy( x, 0, {}, '5' ); // $ExpectError
	flattenFromBy( x, 0, {}, true ); // $ExpectError
	flattenFromBy( x, 0, {}, false ); // $ExpectError
	flattenFromBy( x, 0, {}, null ); // $ExpectError
	flattenFromBy( x, 0, {}, undefined ); // $ExpectError
	flattenFromBy( x, 0, {}, {} ); // $ExpectError
	flattenFromBy( x, 0, {}, [ 1 ] ); // $ExpectError

	flattenFromBy( x, 0, {}, '5', {} ); // $ExpectError
	flattenFromBy( x, 0, {}, true, {} ); // $ExpectError
	flattenFromBy( x, 0, {}, false, {} ); // $ExpectError
	flattenFromBy( x, 0, {}, null, {} ); // $ExpectError
	flattenFromBy( x, 0, {}, undefined, {} ); // $ExpectError
	flattenFromBy( x, 0, {}, {}, {} ); // $ExpectError
	flattenFromBy( x, 0, {}, [ 1 ], {} ); // $ExpectError
}

// The compiler throws an error if the function is provided an unsupported number of arguments...
{
	const x = zeros( 'generic', [ 2, 2, 2 ], 'row-major' );

	flattenFromBy(); // $ExpectError
	flattenFromBy( x ); // $ExpectError
	flattenFromBy( x, 0 ); // $ExpectError
	flattenFromBy( x, 0, {}, identity, {}, {} ); // $ExpectError
}
