/*
* @license Apache-2.0
*
* Copyright (c) 2026 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

import zeros = require( './../../../../zeros' );
import add3 = require( '@stdlib/number/float64/base/add3' );
import ternary = require( './index' );


// TESTS //

// The function returns `undefined`...
{
	const x = zeros( [ 2, 2 ] );
	const y = zeros( [ 2, 2 ] );
	const z = zeros( [ 2, 2 ] );
	const w = zeros( [ 2, 2 ] );
	const arrays = [ x, y, z, w ];

	ternary( arrays, add3 ); // $ExpectType void
}

// The compiler throws an error if the function is provided a first argument which is not an array-like object containing ndarray-like objects...
{
	ternary( 5, add3 ); // $ExpectError
	ternary( true, add3 ); // $ExpectError
	ternary( false, add3 ); // $ExpectError
	ternary( null, add3 ); // $ExpectError
	ternary( undefined, add3 ); // $ExpectError
	ternary( {}, add3 ); // $ExpectError
	ternary( [ 1 ], add3 ); // $ExpectError
	ternary( ( x: number ): number => x, add3 ); // $ExpectError
}

// The compiler throws an error if the function is provided a second argument which is not a ternary function...
{
	const x = zeros( [ 2, 2 ] );
	const y = zeros( [ 2, 2 ] );
	const z = zeros( [ 2, 2 ] );
	const w = zeros( [ 2, 2 ] );
	const arrays = [ x, y, z, w ];

	ternary( arrays, '10' ); // $ExpectError
	ternary( arrays, 5 ); // $ExpectError
	ternary( arrays, true ); // $ExpectError
	ternary( arrays, false ); // $ExpectError
	ternary( arrays, null ); // $ExpectError
	ternary( arrays, undefined ); // $ExpectError
	ternary( arrays, [] ); // $ExpectError
	ternary( arrays, {} ); // $ExpectError
}

// The compiler throws an error if the function is provided an unsupported number of arguments...
{
	const x = zeros( [ 2, 2 ] );
	const y = zeros( [ 2, 2 ] );
	const z = zeros( [ 2, 2 ] );
	const w = zeros( [ 2, 2 ] );
	const arrays = [ x, y, z, w ];

	ternary(); // $ExpectError
	ternary( arrays ); // $ExpectError
	ternary( arrays, add3, 10 ); // $ExpectError
}
